def fetch_fastqs(sample, paired_end, prefix=prefix):

    if paired_end:
        fs = ["{prefix}/data/trim/{sample}_1.fastq", "{prefix}/data/trim/{sample}_2.fastq"]
    else:
        fs = ["{prefix}/data/trim/{sample}.fastq"]

    return expand(fs, sample=sample, prefix=prefix)


def keep_multi_aligning(keep):

    if keep:
        return ""
    else:
        return "grep -E '@|NH:i:1$' |"


rule hisat2:
    input:
        reads=lambda w: fetch_fastqs(w.sample, config["paired_end"], w.prefix),
    output:
        "{prefix}/data/bam/{sample}.bam"
    log:                                # optional
        "{prefix}/data/bam/{sample}.log"
    params:                             # idx is required, extra is optional
        idx=config["hisat2_index_prefix"],
        extra=config["hisat2_extra_flags"]
    threads: 8                          # optional, defaults to 1
    wrapper:
        "0.17.4/bio/hisat2"
